bpf/include/bpf/tailcall.h

```diff
@@ -28,7 +28,7 @@ tail_call_static(const struct __ctx_buff *ctx, const void *map,
 		     "r2 = %[map]\n\t"
 		     "r3 = %[slot]\n\t"
 		     "call 12\n\t"
-		     :: [ctx]"r"(ctx), [map]"r"(map), [slot]"i"(slot)
+		     :: [ctx]"r"(ctx), [map]"r"(map), [slot]"ri"(slot)
```

Comment:
I forgot whether we discussed about this earlier, but does this mean that the whole tail_call_static() does not provide a hard guarantee anymore that the verifier will emit jmp/nops?

Reply:
That's my understanding as well. I'm not sure what that changes in practice given we check `slot` is a compile-time constant above.

Reply:
> That's my understanding as well. I'm not sure what that changes in practice given we check slot is a compile-time constant above.

Did you check JITed code? My assumption is that the compile-time check on the slot might be bypassed. Maybe it makes sense to dup the `tail_call_static()` code into `tail_call_dynamic()` with the `r` as slot, and add a comment that if r is known const we could still get it to work? (That way it wouldn't affect the others.)

Reply:
> 

I think I did when I first wrote that code, but just checked again for good measure:
```
$ sudo bpftool prog dump xlated id 2480 | grep tail_call | wc -l
5
$ sudo bpftool prog dump jited id 2480 | grep lfence | wc -l
1
```
In the above, 2480 is `TEMPLATE_HOST_EP_ID` (i.e., bpf_host's `handle_lxc_traffic`). There are 5 tail calls in the bytecode and only one results in a retpoline. That matches the expectation since we have several `static_tail_call`s but only one `dynamic_tail_call` (line 1282).

If I managed to define `HOST_EP_ID` as a macro (vs. static data currently), I'll also be able to remove this change to the inline asm. That is tracked at https://github.com/cilium/cilium/issues/13121.

Reply:
> If I managed to define HOST_EP_ID as a macro (vs. static data currently), I'll also be able to remove this change to the inline asm.
Ok, worst case we can change `tail_call_dynamic()` as suggested in order to `keep tail_call_static()` only with `i` modifier again.